src/style-spec/expression/definitions/within.js

```diff
@@ -0,0 +1,273 @@
+import type {GeoJSON, GeoJSONPolygon, GeoJSONMultiPolygon} from '@mapbox/geojson-types';
+import MercatorCoordinate from '../../../geo/mercator_coordinate';
+import EXTENT from '../../../data/extent';
+import Point from '@mapbox/point-geometry';
+
+type GeoJSONPolygons =| GeoJSONPolygon | GeoJSONMultiPolygon;
+
+type BBox = [number, number, number, number];
+
+function calcBBox(bbox: BBox, geom, type) {
+    if (type === 'Point') {
+        updateBBox(bbox, geom);
+    } else if (type === 'MultiPoint' || type === 'LineString') {
+        for (let i = 0; i < geom.length; ++i) {
+            updateBBox(bbox, geom[i]);
+        }
+    } else if (type === 'Polygon' || type === 'MultiLineString') {
+        for (let i = 0; i < geom.length; i++) {
+            for (let j = 0; j < geom[i].length; j++) {
+                updateBBox(bbox, geom[i][j]);
+            }
+        }
+    } else if (type === 'MultiPolygon') {
+        for (let i = 0; i < geom.length; i++) {
+            for (let j = 0; j < geom[i].length; j++) {
+                for (let k = 0; k < geom[i][j].length; k++) {
+                    updateBBox(bbox, geom[i][j][k]);
+                }
+            }
+        }
+    }
+}
+
+function updateBBox(bbox: BBox, coord: Point) {
+    bbox[0] = Math.min(bbox[0], coord[0]);
+    bbox[1] = Math.min(bbox[1], coord[1]);
+    bbox[2] = Math.max(bbox[2], coord[0]);
+    bbox[3] = Math.max(bbox[3], coord[1]);
+}
+
+function boxWithinBox(bbox1, bbox2) {
+    if (bbox1[0] <= bbox2[0]) return false;
+    if (bbox1[2] >= bbox2[2]) return false;
+    if (bbox1[1] <= bbox2[1]) return false;
+    if (bbox1[3] >= bbox2[3]) return false;
+    return true;
+}
+
+function getLngLatPoint(coord: Point, canonical) {
+    const tilesAtZoom = Math.pow(2, canonical.z);
+    const x = (coord.x / EXTENT + canonical.x) / tilesAtZoom;
+    const y = (coord.y / EXTENT + canonical.y) / tilesAtZoom;
+    const p = new MercatorCoordinate(x, y).toLngLat();
+
+    return [p.lng, p.lat];
+}
+
+function getLngLatPoints(line, canonical) {
+    const coords = [];
+    for (let i = 0; i < line.length; ++i) {
+        coords.push(getLngLatPoint(line[i], canonical));
+    }
+    return coords;
+}
+
+function onBoundary(p, p1, p2) {
+    const x1 = p[0] - p1[0];
+    const y1 = p[1] - p1[1];
+    const x2 = p[0] - p2[0];
+    const y2 = p[1] - p2[1];
+    return (x1 * y2 - x2 * y1 === 0) && (x1 * x2 <= 0) && (y1 * y2 <= 0);
+}
+
+function rayIntersect(p, p1, p2) {
+    return ((p1[1] > p[1]) !== (p2[1] > p[1])) && (p[0] < (p2[0] - p1[0]) * (p[1] - p1[1]) / (p2[1] - p1[1]) + p1[0]);
+}
+
+// ray casting algorithm for detecting if point is in polygon
+function pointWithinPolygon(point, rings) {
+    let inside = false;
+    for (let i = 0, len = rings.length; i < len; i++) {
+        const ring = rings[i];
+        for (let j = 0, len2 = ring.length; j < len2 - 1; j++) {
+            if (onBoundary(point, ring[j], ring[j + 1])) return false;
```

Comment:
I'm not sure this check is necessary. As far as I know, the general convention for geometric predicates is to not accept objects on the boundary as being contained inside the polygon. See for example an answer to this question https://gis.stackexchange.com/questions/170264/python-point-in-polygon-boundary-and-vertex-check-ray-casting

Reply:
It is necessary, I tried to test the points on the boundary, without this check it would fail.
Referring from https://en.wikipedia.org/wiki/Point_in_polygon:
>  This method won't work if the point is on the edge of the polygon, or if the ray intersects with a corner of the polygon.

Reply:
What I mean is that the point on the boundary shouldn't be considered as "contained within" the polygon according to conventional topology rules. See the articles linked in the answer above:
- https://lin-ear-th-inking.blogspot.com/2007/06/subtleties-of-ogc-covers-spatial.html
- https://shapely.readthedocs.io/en/latest/manual.html#binary-predicates

Reply:
If I understood correctly,  `within` shall return false if the point is on the boundary, so in the code, if a point is on the boundary, it immediately return `false`, it will ignore the following ray casting checks.